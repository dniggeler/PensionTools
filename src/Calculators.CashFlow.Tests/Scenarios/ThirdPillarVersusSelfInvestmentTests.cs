using System.Threading.Tasks;
using Calculators.CashFlow.Models;
using PensionCoach.Tools.CommonTypes;
using PensionCoach.Tools.CommonTypes.Tax;
using Snapshooter.Xunit;
using Xunit;

namespace Calculators.CashFlow.Tests.Scenarios
{
    [Trait("Scenario Calculators", "3a vs Anlagekonto")]
    public class ThirdPillarVersusSelfInvestmentTests : IClassFixture<CashFlowFixture<ITaxScenarioCalculator>>
    {
        private readonly CashFlowFixture<ITaxScenarioCalculator> fixture;

        public ThirdPillarVersusSelfInvestmentTests(CashFlowFixture<ITaxScenarioCalculator> fixture)
        {
            this.fixture = fixture;
        }

        [Fact(DisplayName = "Zero Return on Self-Investment")]
        public async Task Calculate_Zero_Return_On_SelfInvestment()
        {
            // given
            var calculationYear = 2022;
            var finalYear = calculationYear + 10;
            var bfsMunicipalityId = 261;
            TaxPerson person = new()
            {
                Name = "Unit Test Scenario",
                CivilStatus = CivilStatus.Married,
                ReligiousGroupType = ReligiousGroupType.Other,
                PartnerReligiousGroupType = ReligiousGroupType.Other,
                TaxableIncome = 150_000,
                TaxableFederalIncome = 100_000,
                TaxableWealth = 500_000
            };

            ThirdPillarVersusSelfInvestmentScenarioModel scenarioModel = new()
            {
                FinalYear = finalYear,
                InvestmentExcessReturn = 0.03M,
                InvestmentAmount = 7056,
            };

            // when
            var result =
                await fixture.Calculator.ThirdPillarVersusSelfInvestmentAsync(calculationYear, bfsMunicipalityId, person, scenarioModel);

            // then
            Snapshot.Match(result, opt => opt.IgnoreFields("$..Id"));
        }
    }
}
