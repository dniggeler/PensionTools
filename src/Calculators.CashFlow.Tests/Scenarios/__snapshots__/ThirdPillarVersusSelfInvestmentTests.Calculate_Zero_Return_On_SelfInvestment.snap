[
  {
    "State": "IsRight",
    "Right": {
      "StartingYear": 2022,
      "NumberOfPeriods": 11,
      "DeltaSeries": [
        {
          "Year": 2022,
          "Amount": 2299.41,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2023,
          "Amount": 4598.82,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2024,
          "Amount": 6898.2300,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2025,
          "Amount": 9197.640000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2026,
          "Amount": 11497.05000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2027,
          "Amount": 13796.4600000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2028,
          "Amount": 16095.870000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2029,
          "Amount": 18395.28000000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2030,
          "Amount": 20694.6900000000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2031,
          "Amount": 22994.100000000000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2032,
          "Amount": 22847.10000000000000000000,
          "AccountType": "Exogenous"
        }
      ],
      "BenchmarkSeries": [
        {
          "Year": 2022,
          "Amount": -33219.83,
          "AccountType": "Wealth"
        },
        {
          "Year": 2022,
          "Amount": 7056.0,
          "AccountType": "Investment"
        },
        {
          "Year": 2022,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2023,
          "Amount": -66651.34,
          "AccountType": "Wealth"
        },
        {
          "Year": 2023,
          "Amount": 14323.68,
          "AccountType": "Investment"
        },
        {
          "Year": 2023,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2024,
          "Amount": -100300.8804,
          "AccountType": "Wealth"
        },
        {
          "Year": 2024,
          "Amount": 21809.3904,
          "AccountType": "Investment"
        },
        {
          "Year": 2024,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2025,
          "Amount": -134174.992112,
          "AccountType": "Wealth"
        },
        {
          "Year": 2025,
          "Amount": 29519.672112,
          "AccountType": "Investment"
        },
        {
          "Year": 2025,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2026,
          "Amount": -168280.41227536,
          "AccountType": "Wealth"
        },
        {
          "Year": 2026,
          "Amount": 37461.26227536,
          "AccountType": "Investment"
        },
        {
          "Year": 2026,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2027,
          "Amount": -202624.0801436208,
          "AccountType": "Wealth"
        },
        {
          "Year": 2027,
          "Amount": 45641.1001436208,
          "AccountType": "Investment"
        },
        {
          "Year": 2027,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2028,
          "Amount": -237213.143147929424,
          "AccountType": "Wealth"
        },
        {
          "Year": 2028,
          "Amount": 54066.333147929424,
          "AccountType": "Investment"
        },
        {
          "Year": 2028,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2029,
          "Amount": -272054.96314236730672,
          "AccountType": "Wealth"
        },
        {
          "Year": 2029,
          "Amount": 62744.32314236730672,
          "AccountType": "Investment"
        },
        {
          "Year": 2029,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2030,
          "Amount": -307157.1228366383259216,
          "AccountType": "Wealth"
        },
        {
          "Year": 2030,
          "Amount": 71682.6528366383259216,
          "AccountType": "Investment"
        },
        {
          "Year": 2030,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2031,
          "Amount": -342527.432421737475699248,
          "AccountType": "Wealth"
        },
        {
          "Year": 2031,
          "Amount": 80889.132421737475699248,
          "AccountType": "Investment"
        },
        {
          "Year": 2031,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2032,
          "Amount": -352058.10639438959997022544,
          "AccountType": "Wealth"
        },
        {
          "Year": 2032,
          "Amount": 90371.80639438959997022544,
          "AccountType": "Investment"
        },
        {
          "Year": 2032,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        }
      ],
      "ScenarioSeries": [
        {
          "Year": 2022,
          "Amount": -30920.42,
          "AccountType": "Wealth"
        },
        {
          "Year": 2022,
          "Amount": 7056.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2023,
          "Amount": -61840.84,
          "AccountType": "Wealth"
        },
        {
          "Year": 2023,
          "Amount": 14112.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2024,
          "Amount": -92761.26,
          "AccountType": "Wealth"
        },
        {
          "Year": 2024,
          "Amount": 21168.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2025,
          "Amount": -123681.68,
          "AccountType": "Wealth"
        },
        {
          "Year": 2025,
          "Amount": 28224.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2026,
          "Amount": -154602.10,
          "AccountType": "Wealth"
        },
        {
          "Year": 2026,
          "Amount": 35280.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2027,
          "Amount": -185522.52,
          "AccountType": "Wealth"
        },
        {
          "Year": 2027,
          "Amount": 42336.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2028,
          "Amount": -216442.94,
          "AccountType": "Wealth"
        },
        {
          "Year": 2028,
          "Amount": 49392.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2029,
          "Amount": -247363.36,
          "AccountType": "Wealth"
        },
        {
          "Year": 2029,
          "Amount": 56448.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2030,
          "Amount": -278283.78,
          "AccountType": "Wealth"
        },
        {
          "Year": 2030,
          "Amount": 63504.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2031,
          "Amount": -309204.20,
          "AccountType": "Wealth"
        },
        {
          "Year": 2031,
          "Amount": 70560.0,
          "AccountType": "ThirdPillar"
        },
        {
          "Year": 2032,
          "Amount": -238839.200,
          "AccountType": "Wealth"
        },
        {
          "Year": 2032,
          "Amount": 0.0,
          "AccountType": "ThirdPillar"
        }
      ]
    },
    "Left": null
  }
]
